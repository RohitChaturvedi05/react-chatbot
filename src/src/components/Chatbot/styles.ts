import styled, { css } from "styled-components";

const shadow = css`
  box-shadow:
    0 4px 8px rgba(0, 0, 0, 0.2),
    0 6px 20px rgba(0, 0, 0, 0.19);
`;
export const ChatContainer = styled.div<{
  height: string;
  width: string;
  $noShadow?: boolean;
}>`
  display: flex;
  flex-direction: column;
  border: 1px solid ${(props) => props.theme.pallet.divider};
  height: ${({ height }) => height};
  width: ${({ width }) => width};
  ${({ $noShadow }) => !$noShadow && shadow}
`;
