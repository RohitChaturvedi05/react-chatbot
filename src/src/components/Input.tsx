import SendIcon from "@mui/icons-material/Send";
import { useState } from "react";
import styled from "styled-components";
import { InputProps } from "../types";
import { Button } from "./Button";

const Footer = styled.div`
  padding: ${(props) => props.theme.spacing.sm};
  background-color: ${(props) => props.theme.pallet.background.dark};
`;
const InputContainer = styled.div`
  display: flex;
  flex-direction: row;
  border-radius: ${(props) => props.theme.spacing.sm};
  gap: ${(props) => props.theme.spacing.sm};
  input {
    width: 100%;
    padding: ${(props) => props.theme.spacing.sm};
    border: none;
    background-color: ${(props) => props.theme.pallet.background.main};
    color: ${(props) => props.theme.pallet.background.text};
  }
`;

export const Input = ({
  icon,
  placeholder = "Type your message...",
  buttonLabel = "Send",
  onSubmit,
  onKeyDown,
  ...rest
}: InputProps) => {
  const [value, setValue] = useState("");

  const handleSubmit = () => {
    if (value.trim()) {
      onSubmit(value);
      setValue("");
    }
  };

  return (
    <Footer>
      <InputContainer>
        <input
          placeholder={placeholder}
          value={value}
          onChange={(e) => setValue(e.target.value)}
          onKeyDown={(e) => {
            if (e.key === "Enter") {
              handleSubmit();
            }
            if (onKeyDown) onKeyDown(e);
          }}
          {...rest}
        />
        <Button
          $type="primary"
          $displayStyle="outline"
          onClick={handleSubmit}
          aria-label="Send message"
        >
          {buttonLabel} {icon ?? <SendIcon />}
        </Button>
      </InputContainer>
    </Footer>
  );
};
