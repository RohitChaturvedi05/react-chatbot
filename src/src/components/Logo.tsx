import styled from "styled-components";
import { theme } from "../theme";

export const LogoStyled = styled.img`
  width: 40px;
  height: 40px;
  border-radius: ${theme.radius.circle};
  object-fit: cover;
`;

export const Logo = (props: React.ImgHTMLAttributes<HTMLImageElement>) => (
  <LogoStyled {...props} />
);
