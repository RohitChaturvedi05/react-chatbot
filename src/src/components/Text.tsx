import Typography, { TypographyProps } from "@mui/material/Typography";
import styled from "styled-components";

const TypographyStyled = styled(Typography)`
  [dir="rtl"] & {
    text-align: right;
  }
`;

export type TextProps = TypographyProps & {};

export const Text = ({ ...props }: TextProps) => (
  <TypographyStyled fontWeight="lighter" {...props} />
);
