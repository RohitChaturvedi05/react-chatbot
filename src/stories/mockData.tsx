import { Messages } from "../src/types";

export const MockChats: Messages = [
  {
    id: 1,
    role: "assistant",
    content: "Welcome to the chatbot! How can I help you today?",
  },
  {
    id: 2,
    role: "user",
    content:
      "Hi! I need some help with my order. I tried checking the status online, but I couldn't find the information I was looking for. Could you please assist me with finding the details?",
  },
  {
    id: 3,
    role: "assistant",
    content:
      "Sure, I can assist you with that. Can you please provide your order number so I can pull up the relevant details for you? Once I have your order number, I can check the status, delivery date, and any other information you might need regarding your order.",
  },
  {
    id: 4,
    role: "user",
    content: "My order number is 123456.",
  },
  {
    id: 5,
    role: "assistant",
    content:
      "Thank you! Let me check the details for your order. It will just take a moment. While I'm looking into this, is there anything specific you're concerned about with your order, like delivery timing or the condition of the items? Your input will help me assist you better.",
  },
  {
    id: 6,
    role: "assistant",
    content:
      "I see that your order is scheduled for delivery tomorrow. Everything seems to be on track. The package is currently in transit and should arrive by the end of the day tomorrow. Is there anything else you need help with, or is there anything else you would like to change about your order?",
  },
  {
    id: 7,
    role: "user",
    content: "Can I reschedule the delivery?",
  },
  {
    id: 8,
    role: "assistant",
    content:
      "Yes, you can reschedule the delivery. Would you like me to do that for you now, or would you prefer to choose a specific date and time that works best for you? Just let me know what your preference is, and I'll make sure the delivery is adjusted accordingly.",
  },
  {
    id: 9,
    role: "user",
    content: "Yes, please reschedule it for Friday.",
  },
  {
    id: 10,
    role: "assistant",
    content:
      "Your delivery has been successfully rescheduled for Friday. You will receive a confirmation email shortly with the updated delivery details. If you have any other questions or need further assistance with anything else, feel free to ask. I'm here to help make sure everything goes smoothly for you.",
  },
  // {
  //   id: 11,
  //   title: "Quiz",
  //   role: "assistant",
  //   content: {
  //     widget: "quiz",
  //     question: "What are the benefits of using TypeScript over JavaScript?",
  //     options: [
  //       {
  //         id: "11",
  //         text: "TypeScript provides static typing which helps catch errors early",
  //       },
  //       { id: "12", text: "TypeScript compiles directly into machine code" },
  //       { id: "13", text: "TypeScript allows for dynamic typing at runtime" },
  //       {
  //         id: "14",
  //         text: "TypeScript can be run directly in the browser without compilation",
  //       },
  //     ],
  //   },
  // },
  // {
  //   id: 12,
  //   title: "List",
  //   role: "assistant",
  //   content: {
  //     widget: "list",
  //     title: "Top Programming Languages",
  //     items: [
  //       "JavaScript",
  //       "Python",
  //       "Java",
  //       "C++",
  //       "C#",
  //       "Ruby",
  //       "Swift",
  //       "Go",
  //       "PHP",
  //       "Rust",
  //     ],
  //   },
  // },
];
