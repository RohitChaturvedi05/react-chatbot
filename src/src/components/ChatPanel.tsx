import { ReactElement } from "react";
import styled from "styled-components";
import { theme } from "../theme";
import { Header } from "./Header/Header";
import { Input } from "./Input";
import { Messages } from "./Messages";

type AllowedChildren =
  | ReactElement<typeof Header>
  | ReactElement<typeof Messages>
  | ReactElement<typeof Input>;

const ChatContainer = styled.div`
  height: 100vh;
  display: flex;
  flex-direction: column;
  border: 1px solid ${theme.colors.divider};
  border-radius: ${theme.radius.sm};
`;

export type ChatPanelProps = {
  children: AllowedChildren | AllowedChildren[] | React.ReactNode[];
};

export const ChatPanel = ({ children }: ChatPanelProps) => {
  return <ChatContainer>{children}</ChatContainer>;
};

ChatPanel.Header = Header;
ChatPanel.Messages = Messages;
ChatPanel.Input = Input;
