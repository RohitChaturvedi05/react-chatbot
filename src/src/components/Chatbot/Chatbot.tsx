import { ThemeProvider } from "styled-components";
import { NOOP } from "../../constants";
import { ChatProvider } from "../../hooks/useChatContext";
import { useOutsideClick } from "../../hooks/useOutsideClick";
import { createTheme } from "../../theme";
import { ChatbotProps } from "../../types";
import { Header } from "../Header/Header";
import { Input } from "../Input";
import { Messages } from "../Messages";
import { ChatContainer } from "./styles";

export const Chatbot = ({
  height = "100vh",
  width = "auto",
  noShadow,
  children,
  theme = "light",
  onClose = NOOP,
  onOutSideClick = NOOP,
  ...rest
}: ChatbotProps) => {
  const ref = useOutsideClick(onOutSideClick);
  const chatTheme = createTheme(theme);
  return (
    <ThemeProvider theme={chatTheme}>
      <ChatProvider value={{ onOutSideClick, onClose }}>
        <ChatContainer
          ref={ref}
          $noShadow={noShadow}
          height={height}
          width={width}
          {...rest}
        >
          {children}
        </ChatContainer>
      </ChatProvider>
    </ThemeProvider>
  );
};

Chatbot.Header = Header;
Chatbot.Messages = Messages;
Chatbot.Input = Input;
