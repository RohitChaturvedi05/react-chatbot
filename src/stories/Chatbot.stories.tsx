import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import { Chatbot } from "../src/components/Chatbot";
import { ChatbotExample, DrawerExample, Example } from "./Examples";

const meta: Meta<typeof Chatbot> = {
  title: "Chatbot",
  component: Chatbot,
  argTypes: {
    height: {
      control: { type: "text" },
    },
    width: {
      control: { type: "text" },
    },
    noShadow: {
      control: { type: "boolean" },
    },
    theme: {
      control: { disable: true },
    },
    children: {
      control: { disable: true },
    },
  },
};

type Story = StoryObj<typeof Chatbot>;

export const Playground: Story = {
  render: (args, context) => {
    const theme = context.globals.theme;
    return <ChatbotExample {...args} theme={theme} />;
  },
};

export const PopupExample: Story = {
  render: (args, context) => {
    const theme = context.globals.theme;
    return <Example {...args} theme={theme} />;
  },
};
export const DrawerExamples: Story = {
  render: (args, context) => {
    const theme = context.globals.theme;
    return <DrawerExample {...args} theme={theme} />;
  },
};

export default meta;
