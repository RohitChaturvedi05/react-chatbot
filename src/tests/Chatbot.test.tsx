import { render } from "@testing-library/react";
import React from "react";
import { describe, expect, it, vi } from "vitest";
import { Chatbot } from "../src/components/Chatbot";
import { Logo } from "../src/components/Logo";
import { MockChats } from "../stories/mockData";

vi.mock("../src/hooks/useAutoScroll", () => ({ useAutoScroll: vi.fn() }));

describe("Chatbot", () => {
  it("renders correctly with children", () => {
    const { asFragment } = render(
      <Chatbot>
        <Chatbot.Header
          title="React Chatbot"
          logo={<Logo src="/chatbot-male.png" alt="chatbot-male" />}
        />
        <Chatbot.Messages
          messages={MockChats}
          maxWidth="70%"
          onContentAction={console.log}
        />
        <Chatbot.Input onSubmit={console.log} />
      </Chatbot>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
