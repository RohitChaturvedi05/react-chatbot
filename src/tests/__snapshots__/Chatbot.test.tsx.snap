// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chatbot > renders correctly with children 1`] = `
<DocumentFragment>
  <div
    class="sc-dpBQxM oHChs"
    height="100vh"
    width="auto"
  >
    <header
      class="sc-egkSDF dzOdls"
    >
      <div
        class="sc-fAUdSK fsTGZU"
      >
        <img
          alt="chatbot-male"
          class="sc-jwIPbr bgGhze"
          src="/chatbot-male.png"
        />
      </div>
      <p
        class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-kbdbbi-MuiTypography-root"
      >
        React Chatbot
      </p>
      <button
        class="sc-blHHSb bnRRnm"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  css-20bmp1-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </header>
    <div
      class="sc-ixGGxD jpaJcb"
    >
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Welcome to the chatbot! How can I help you today?
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu icVAgW"
        role="user"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb fsyWtk"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Hi! I need some help with my order. I tried checking the status online, but I couldn't find the information I was looking for. Could you please assist me with finding the details?
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Sure, I can assist you with that. Can you please provide your order number so I can pull up the relevant details for you? Once I have your order number, I can check the status, delivery date, and any other information you might need regarding your order.
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu icVAgW"
        role="user"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb fsyWtk"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            My order number is 123456.
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Thank you! Let me check the details for your order. It will just take a moment. While I'm looking into this, is there anything specific you're concerned about with your order, like delivery timing or the condition of the items? Your input will help me assist you better.
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            I see that your order is scheduled for delivery tomorrow. Everything seems to be on track. The package is currently in transit and should arrive by the end of the day tomorrow. Is there anything else you need help with, or is there anything else you would like to change about your order?
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu icVAgW"
        role="user"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb fsyWtk"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Can I reschedule the delivery?
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Yes, you can reschedule the delivery. Would you like me to do that for you now, or would you prefer to choose a specific date and time that works best for you? Just let me know what your preference is, and I'll make sure the delivery is adjusted accordingly.
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu icVAgW"
        role="user"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb fsyWtk"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Yes, please reschedule it for Friday.
          </p>
        </div>
      </div>
      <div
        class="sc-kLhKbu iAjnur"
        role="assistant"
      >
        <div
          class="sc-dstKZu HWmsM"
        />
        <div
          class="sc-khLCKb qDrqh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1  sc-gtLWhw kLKHwq css-mwd914-MuiTypography-root"
          >
            Your delivery has been successfully rescheduled for Friday. You will receive a confirmation email shortly with the updated delivery details. If you have any other questions or need further assistance with anything else, feel free to ask. I'm here to help make sure everything goes smoothly for you.
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-ivxoEo dpoXUW"
    >
      <div
        class="sc-ghWlax iyHqAz"
      >
        <input
          placeholder="Type your message..."
          value=""
        />
        <button
          aria-label="Send message"
          class="sc-blHHSb ehMxOH"
        >
          Send 
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium  css-20bmp1-MuiSvgIcon-root"
            data-testid="SendIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
